### PR TITLE
(1155) Fractionalization based on ERC20 / only on L1 

### PR DESCRIPTION
**Pull request by elmariachi111**
- cherry picked a lot of history from #57 
- dropped all the L2 related code
- replaces a shared ERC1155 impl with ERC20 
- initializes ERC20 with derived names & symbols
- kept the metadata code for the Fractionalizer even though it's not a token itself anymore
- uses minimal proxy clones to save maximum gas
- merged and fixed all relevant tests that worked so far
- bound the Fractionalizer contract directly to IPNFT (might allow NFT locking)
- kept the share claiming code with mandatory terms sigs 
- reduced the subgraphs and merged the handlers. This must be consolidated.

### CODE REVIEW

File: test/helpers/MakeGnosisWallet.sol [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import "forge-std/console.sol";
+import "forge-std/Test.sol";
+
+import { GnosisSafeL2 } from "safe-global/safe-contracts/GnosisSafeL2.sol";
+import { GnosisSafeProxyFactory } from "safe-global/safe-contracts/proxies/GnosisSafeProxyFactory.sol";
+import { Enum } from "safe-global/safe-contracts/common/Enum.sol";
+import { DefaultCallbackHandler } from "safe-global/safe-contracts/handler/DefaultCallbackHandler.sol";
+//import { CompatibilityFallbackHandler } from "safe-global/safe-contracts/handler/CompatibilityFallbackHandler.sol";
+
+library MakeGnosisWallet {
```
Comment (firattale) [inline test/helpers/MakeGnosisWallet.sol:13]: should we remove this lib?

File: src/SchmackoSwap.sol [hunk 2]
```diff
@@ -64,14 +71,18 @@ contract SchmackoSwap is ERC165, ReentrancyGuard {
     /// @param creator The address of the seller
     /// @param askPrice The amount the seller is asking for in exchange for the token
     struct Listing {
-        ERC1155Supply tokenContract;
+        IERC1155Supply tokenContract;
```
Comment (firattale) [inline src/SchmackoSwap.sol:74]: why do we use ERC1155 here?

File: test/helpers/MakeGnosisWallet.sol (hunk 1 above)
Comment (elmariachi111) [reply test/helpers/MakeGnosisWallet.sol:13]: why would you? it's super helpful to setup a gnosis wallet for tests...

File: src/SchmackoSwap.sol (hunk 2 above)
Comment (elmariachi111) [reply src/SchmackoSwap.sol:74]: Schmackoswap is a generic IERC1155 sales contract.

File: src/Fractionalizer.sol [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+     * @param fractionsAmount  uint256  the initial amount of fractions issued
+     * @param agreementCid    bytes32  a content hash that identifies the terms underlying the issued fractions
+     */
+
+    function fractionalizeIpnft(uint256 ipnftId, uint256 fractionsAmount, string calldata agreementCid) external returns (uint256 fractionId) {
+        if (ipnft.totalSupply(ipnftId) != 1) {
+            revert("IPNFT supply must be 1");
+        }
+        if (ipnft.balanceOf(_msgSender(), ipnftId) != 1) {
+            revert("only owner can initialize fractions");
```
Comment (firattale) [inline src/Fractionalizer.sol:123]: can we use Custom Errors to save some gas

File: src/Fractionalizer.sol [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+
+        // ensure we can only call this once per sales cycle
+        if (address(fractionalized[fractionId].originalOwner) != address(0)) {
+            revert("token is already fractionalized");
+        }
+
+        // https://github.com/OpenZeppelin/workshops/tree/master/02-contracts-clone
+        FractionalizedTokenUpgradeable fractionalizedToken = FractionalizedTokenUpgradeable(Clones.clone(tokenImplementation));
+        string memory name = string(abi.encodePacked("Fractions of IPNFT #", Strings.toString(ipnftId)));
+        string memory symbol = string(string(abi.encodePacked(ipnftSymbol, "-FAM")));
```
Comment (firattale) [inline src/Fractionalizer.sol:140]: why do we need additional string here?

File: src/Fractionalizer.sol (hunk 3 above)
Comment (elmariachi111) [reply src/Fractionalizer.sol:123]: sure, but they're far easier to write as long as you're in coding mode and they're quite hard to resolve on the client side. They mostly save gas during deployment, not during execution (only when the tx reverts), but I'm happily converting them, as we did on IPNFT, in the last roundtrip.

File: src/Fractionalizer.sol (hunk 4 above)
Comment (elmariachi111) [reply src/Fractionalizer.sol:140]: well, that's imo a requirement? The IPNFT has a symbol and its fractions have a symbol that derives from it.

**Review by firattale — APPROVED** [review]

**Pull request closed without merging**